zipkin-cassandra/src/main/scala/com/twitter/zipkin/storage/cassandra/CassandraSpanStore.scala

```diff
@@ -94,91 +93,106 @@ class CassandraSpanStore(
-        .filter { a => !Constants.CoreAnnotations.contains(a.value) }
-        .groupBy(_.value)
-        .foreach { case (_, as) =>
-          val a = as.min
-          a.host foreach { endpoint =>
-            IndexAnnotationCounter.incr()
-
-            repository.storeTraceIdByAnnotation(
-              annotationKey(endpoint.serviceName, a.value, None), timestamp, span.traceId, indexTtl.inSeconds)
-          }
-        }
-
-      span.binaryAnnotations foreach { ba =>
-        ba.host foreach { endpoint =>
+      val annotationsFuture = Future.join(
+        span.annotations
+          .filter { a => !Constants.CoreAnnotations.contains(a.value) }
+          .groupBy(_.value)
+          .flatMap { case (_, as) =>
+            val a = as.min
+            a.host map { endpoint =>
+              IndexAnnotationCounter.incr()
+
+              FutureUtil.toFuture(
+                repository.storeTraceIdByAnnotation(
+                  annotationKey(endpoint.serviceName, a.value, None), timestamp, span.traceId, indexTtl.inSeconds))
+            }
+          }.toList)
+
+      val binaryFuture = Future.join(span.binaryAnnotations flatMap { ba =>
+        ba.host map { endpoint =>
           IndexBinaryAnnotationCounter.incr()
 
-          repository.storeTraceIdByAnnotation(
-            annotationKey(endpoint.serviceName, ba.key, Some(ba.value)), timestamp, span.traceId, indexTtl.inSeconds)
-
-          repository.storeTraceIdByAnnotation(
-            annotationKey(endpoint.serviceName, ba.key, None), timestamp, span.traceId, indexTtl.inSeconds)
+          Future.join(
+            FutureUtil.toFuture(
+              repository.storeTraceIdByAnnotation(
+                annotationKey(endpoint.serviceName, ba.key, Some(ba.value)), timestamp, span.traceId, indexTtl.inSeconds)),
+            FutureUtil.toFuture(
+              repository.storeTraceIdByAnnotation(
+                annotationKey(endpoint.serviceName, ba.key, None), timestamp, span.traceId, indexTtl.inSeconds)))
         }
-      }
-    }
+      })
+
+      Future.join(annotationsFuture, binaryFuture).map(_ => ())
+    } getOrElse Future.value(())
   }
 
   private[this] def getSpansByTraceIds(traceIds: Seq[Long], count: Int): Future[Seq[Seq[Span]]] = {
-    pool {
-      val spans = repository.getSpansByTraceIds(traceIds.toArray.map(Long.box), count)
-        .mapValues { case spans :java.util.List[ByteBuffer] => spans.asScala.map(spanCodec.decode(_).toSpan) }
+    FutureUtil.toFuture(repository.getSpansByTraceIds(traceIds.toArray.map(Long.box), count))
+      .map { spansByTraceId =>
+        val spans =
+          spansByTraceId.asScala.mapValues { spans => spans.asScala.map(spanCodec.decode(_).toSpan) }
 
-      traceIds.map(traceId => spans.get(traceId)).flatten
-        .sortBy(_.head) // CQL doesn't allow order by with an "in" query
-    }
+        traceIds.flatMap(traceId => spans.get(traceId))
+          .sortBy(_.head) // CQL doesn't allow order by with an "in" query
```

Comment:
not related to this PR, but is this trailing comment still applicable after https://github.com/openzipkin/zipkin/pull/742 got merged?